discovery/dockerswarm/dockerswarm.go

```diff
@@ -0,0 +1,160 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dockerswarm
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/go-kit/kit/log"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/util/strutil"
+)
+
+const (
+	swarmLabel                        = model.MetaLabelPrefix + "dockerswarm_"
+	swarmLabelEndpointPortName        = swarmLabel + "endpoint_port_name"
+	swarmLabelEndpointPortPublishMode = swarmLabel + "endpoint_port_publish_mode"
+	swarmLabelServiceID               = swarmLabel + "service_id"
+	swarmLabelServiceLabelPrefix      = swarmLabel + "service_label_"
+	swarmLabelServiceName             = swarmLabel + "service_name"
+	swarmLabelServiceUpdatingStatus   = swarmLabel + "service_updating_status"
+	swarmLabelVirtualIPAddr           = swarmLabel + "virtual_ip_address"
+	swarmLabelVirtualIPNetworkID      = swarmLabel + "virtual_ip_network_id"
+)
+
+// DefaultSDConfig is the default Docker Swarm SD configuration.
+var DefaultSDConfig = SDConfig{
+	RefreshInterval: model.Duration(60 * time.Second),
+}
+
+// SDConfig is the configuration for Docker Swarm based service discovery.
+type SDConfig struct {
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+
+	Host     string `yaml:"host"`
+	Insecure bool   `yaml:"insecure"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Discovery periodically performs Docker Swarm requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client *client.Client
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "dockerswarm_sd", false)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []client.Opt{
+		client.WithHost(conf.Host),
+		client.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   5 * time.Duration(conf.RefreshInterval),
```

Comment:
That seems a bit long

Reply:
Unless users start lowering the refresh interval to 10s. WDYT? We do this for other SD's.

Reply:
Having a timeout longer than the refresh doesn't seem right, what other SDs do this?

Reply:
openstack

Reply:
solved